src/machines/commerce-machine.ts

```diff
@@ -145,43 +173,26 @@ export const commerceMachine = createMachine<
       },
     },
     actions: {
+      switchPriceIdAndAssociatedData: assign((context, event) => {
+        if (event.type !== 'SWITCH_PRICE') return {}
+
+        const currentPlan = context.pricingData.plans.find(
+          (plan) => plan.stripe_price_id === event.priceId,
+        )
+        const noCouponToApply = !currentPlan?.price_discounted
+        const couponData = noCouponToApply
+          ? {couponToApply: undefined}
+          : extractAppliedDefaultCoupon(context.pricingData)
+
+        return {
+          priceId: event.priceId,
+          ...couponData,
+        }
```

Comment:
Here is the bulk of the fix: we check if a couple should be applied (based on the server-computed `price_discounted` value) and update the machine context accordingly. For site-wide coupons that have a minimum, we'll detect whether it applies to the specific plan and either apply it or wipe it out with `undefined`.